### PR TITLE
opentelemetry-collector: add NixOS module

### PR DESCRIPTION
**Pull request by bouk**
###### Description of changes

Also: like we set it in opentelemetry-collector-contrib, we should set the mainProgram for opentelemetry-collector. This makes the module work for both packages.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/opentelemetry-collector.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+let
+  inherit (lib) mkEnableOption mkIf mkOption types getExe;
+
+  cfg = config.services.opentelemetry-collector;
+  opentelemetry-collector = cfg.package;
+
+  settingsFormat = pkgs.formats.yaml {};
+in {
+  options.services.opentelemetry-collector = {
+    enable = mkEnableOption (lib.mdDoc "Opentelemetry Collector");
```
Comment (RaitoBezarius) [inline nixos/modules/services/opentelemetry-collector.nix:12]: nit:
```suggestion
    enable = mkEnableOption (lib.mdDoc "OpenTelemetry Collector");
```

File: nixos/modules/services/opentelemetry-collector.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      default = pkgs.opentelemetry-collector;
+      defaultText = lib.literalExpression "pkgs.opentelemetry-collector";
+      description = lib.mdDoc "The opentelemetry-collector package to use.";
+    };
+
+    settings = mkOption {
+      type = settingsFormat.type;
+      default = {};
+      description = lib.mdDoc ''
+        Specify the configuration for Opentelemetry Collector in Nix.
```
Comment (RaitoBezarius) [inline nixos/modules/services/opentelemetry-collector.nix:25]: nit:
```suggestion
        Specify the configuration for OpenTelemetry Collector in Nix.
```

File: nixos/modules/services/opentelemetry-collector.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        See https://opentelemetry.io/docs/collector/configuration/ for available options.
+      '';
+    };
+
+    configFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = lib.mdDoc ''
+        Specify a path to a configuration file that Opentelemetry Collector should use.
```
Comment (RaitoBezarius) [inline nixos/modules/services/opentelemetry-collector.nix:35]: nit:
```suggestion
        Specify a path to a configuration file that OpentTelemetry Collector should use.
```

File: nixos/modules/services/opentelemetry-collector.nix [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      );
+      message  = ''
+        Please specify a configuration for Opentelemetry Collector with either
+        'services.opentelemetry-collector.settings' or
+        'services.opentelemetry-collector.configFile'.
+      '';
+    }];
+
+    systemd.services.opentelemetry-collector = {
+      description = "Opentelemetry Collector Service Daemon";
```
Comment (RaitoBezarius) [inline nixos/modules/services/opentelemetry-collector.nix:53]: nit:
```suggestion
      description = "OpenTelemetry Collector Service Daemon";
```

**Review by RaitoBezarius — APPROVED** [review]
This looks very fine, I would appreciate though a NixOS test, do you think it is possible?

**Comment by bouk** [comment]
I've fixed the module naming and I've squashed.

I can look into creating a test, I've never done it before so it might take some time I'm afraid.

**Comment by tylerjl** [comment]
Thank you for preparing this module @bouk - I took the liberty of writing a nixosTest for this in my fork, I branched off your branch [here](https://github.com/bouk/nixpkgs/compare/bouk/mainpr...tylerjl:otel-nixos-test?expand=1) to add the test. If you'd like to just merge/cherry pick into your branch to update this PR, I think that'll adequately cover the module :+1:

**Comment by bouk** [comment]
Thank you @domenkozar

**Pull request merged**